doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,348 @@
+| _absent_   | _default behavior_                                                                                                  |
+| `detached` | The new process is not attached to a console session (similar to `DETACHED_PROCESS`) unless one was inherited.      |
+
+An application that specifies the `detached` allocation policy will _not_ present a console window when launched by
+Explorer, Task Scheduler, etc.
+
+### Interaction with existing APIs
+
+[`CreateProcess`] supports a number of [process creation flags] that dictate how a spawned application will behave with
+regards to console allocation:
+
+* `DETACHED_PROCESS`: No console inheritance, no console host spawned for the new process.
+* `CREATE_NEW_CONSOLE`: No console inheritance, new console host **is** spawned for the new process.
+* `CREATE_NO_WINDOW`: No console inheritance, new console host **is** spawned for the new process.
+    * this is the same as `CREATE_NEW_CONSOLE`, except that the first connection packet specifies that the window should
+      be invisible
+
+Due to the design of [`CreateProcess`] and `ShellExecute`, this specification recommends that an allocation policy of
+`detached` _override_ the inclusion of `CREATE_NEW_CONSOLE` in the `dwFlags` parameter to [`CreateProcess`].
+
+> **Note**
+> `ShellExecute` passes `CREATE_NEW_CONSOLE` _by default_ on all invocations. This impacts our ability to resolve the
+> conflicts between these two APIs--`detached` policy and `CREATE_NEW_CONSOLE`--without auditing every call site in
+> every Windows application that calls `ShellExecute` on a console application. Doing so is infeasible.
+
+### Application impact
+
+An application that opts into the `detached` console allocation policy will **not** be allocated a console unless one is
+inherited. This presents an issue for applications like PowerShell that do want a console window when they are launched
+directly.
+
+Applications in this category can call `AllocConsole()` early in their startup to get fine-grained control over when a
+console is presented.
+
+The call to `AllocConsole()` will fail safely if the application has already inherited a console handle. It will succeed
+if the application does not currently have a console handle.
+
+> **Note**
+> **Backwards Compatibility**: The behavior of `AllocConsole()` is not changing in response to this specification;
+> therefore, applications that intend to run on older versions of Windows that do not support console allocation
+> policies, which call `AllocConsole()`, will continue to behave normally.
+
+### New APIs
+
+Because a console-subsystem application may still want fine-grained control over when and how its console window is
+spawned, we propose the inclusion of a new API, `AllocConsoleEx(PCONSOLE_ALLOCATE_INFO)`.
+
+#### `AllocConsoleEx`
+
+```c++
+typedef
+enum _CONSOLE_ALLOCATE_MODE
+{
+    CONSOLE_ALLOCATE_DEFAULT = 0,
+    CONSOLE_ALLOCATE_HIDDEN,
+    CONSOLE_ALLOCATE_MINIMIZED
+} CONSOLE_ALLOCATE_MODE;

```

Comment:
Should we replace this with a plain `DWORD` and just say that any `ShowWindow` flag can be used here?

Reply:
They're not exactly ShowWindow flags - minimized vs hidden get routed via different paths. we would need to resolve the ShowWindow flag inside conclnt if that's the case.